File: src/ruleset/evaluate.js
/**
 * First-match classification ruleset evaluation.
 *
 * Outcome order is semantic: the first matching `outcomes[]` entry wins;
 * otherwise `default_outcome` applies. Do not reorder ruleset YAML casually.
 *
 * Effective-date precedence: outcome.applicable_from overrides each
 * activated requirement's applicable_from (which records the earliest
 * pathway). Same articles can bind on different dates by classification
 * (e.g. Annex III 2027-12-02 vs Annex I 2028-08-02).
 */

/**
 * @param {unknown} expected
 * @param {unknown} actual
 */
function answerMatches(expected, actual) {
  if (Array.isArray(expected)) {
    return expected.includes(actual);
  }
  if (typeof expected === "boolean") {
    return Boolean(actual) === expected;
  }
  return actual === expected;
}

/**
 * @param {Record<string, unknown>} when
 * @param {Record<string, unknown>} answers
 */
export function matchesWhen(when, answers) {
  for (const [qid, expected] of Object.entries(when || {})) {
    if (!answerMatches(expected, answers[qid])) return false;
  }
  return true;
}

/**
 * Pick the first matching outcome, or the default.
 *
 * @param {object} ruleset
 * @param {Record<string, unknown>} answers
 */
export function selectOutcome(ruleset, answers) {
  for (const outcome of ruleset.outcomes || []) {
    if (matchesWhen(outcome.when, answers)) {
      return { outcome, matched: "outcome", index: ruleset.outcomes.indexOf(outcome) };
    }
  }
  return { outcome: ruleset.default_outcome, matched: "default", index: null };
}

/**
 * Resolve activated obligation ids + effective dates for an answer set.
 *
 * @param {object} options
 * @param {object} options.ruleset Classification ruleset doc
 * @param {Record<string, unknown>} options.answers question id -> answer
 * @param {Map<string, object>|Record<string, object>} [options.requirementsById]
 *   Optional corpus requirements keyed by full id (`framework/version/ref`)
 *   for title + requirement-level applicable_from lookup.
 */
export function evaluateRuleset({ ruleset, answers, requirementsById = {} }) {
  const reqMap =
    requirementsById instanceof Map
      ? requirementsById
      : new Map(Object.entries(requirementsById));

  const { outcome, matched, index } = selectOutcome(ruleset, answers);
  if (!outcome) {
    throw new Error("Ruleset has no matching outcome and no default_outcome");
  }

  const framework = ruleset.framework;
  const outcomeDate = outcome.applicable_from ?? null;
  const activates = outcome.activates || [];

  const obligations = activates.map((ref) => {
    const id = `${framework}/${ref}`;
    const req = reqMap.get(id);
    const requirementDate = req?.applicable_from ?? null;
    // Outcome date wins when present; else fall back to requirement date.
    const applicableFrom = outcomeDate ?? requirementDate;
    return {
      ref,
      id,
      title: req?.title ?? null,
      applicable_from: applicableFrom,
      requirement_applicable_from: requirementDate,
      date_source: outcomeDate != null ? "outcome" : "requirement",
    };
  });

  return {
    ruleset: ruleset.ruleset,
    framework,
    classification: outcome.classification,
    matched,
    outcome_index: index,
    note: outcome.note ?? null,
    outcome_applicable_from: outcomeDate,
    activates: obligations,
  };
}

File: lib/isomer/ruleset/evaluate.ex
defmodule Isomer.Ruleset.Evaluate do
  @moduledoc """
  First-match classification ruleset evaluation.

  Outcome order is semantic. Outcome `applicable_from` overrides requirement-level dates.
  """

  def evaluate(ruleset, answers, requirements_by_id \\ %{}) do
    {outcome, matched, index} = select_outcome(ruleset, answers)

    if is_nil(outcome) do
      raise ArgumentError, "Ruleset has no matching outcome and no default_outcome"
    end

    framework = ruleset["framework"]
    outcome_date = outcome["applicable_from"]
    activates = outcome["activates"] || []

    obligations =
      Enum.map(activates, fn ref ->
        id = "#{framework}/#{ref}"
        req = Map.get(requirements_by_id, id, %{})
        requirement_date = req["applicable_from"]
        applicable_from = outcome_date || requirement_date

        %{
          "ref" => ref,
          "id" => id,
          "title" => req["title"],
          "applicable_from" => applicable_from,
          "requirement_applicable_from" => requirement_date,
          "date_source" => if(outcome_date, do: "outcome", else: "requirement")
        }
      end)

    %{
      "ruleset" => ruleset["ruleset"],
      "framework" => framework,
      "classification" => outcome["classification"],
      "matched" => matched,
      "outcome_index" => index,
      "note" => outcome["note"],
      "outcome_applicable_from" => outcome_date,
      "activates" => obligations
    }
  end

  def select_outcome(ruleset, answers) do
    outcomes = ruleset["outcomes"] || []

    case Enum.find_index(outcomes, &matches_when?(&1["when"], answers)) do
      nil ->
        {ruleset["default_outcome"], "default", nil}

      idx ->
        {Enum.at(outcomes, idx), "outcome", idx}
    end
  end

  def matches_when?(when_map, answers) do
    Enum.all?(when_map || %{}, fn {qid, expected} ->
      answer_matches?(expected, Map.get(answers, qid) || Map.get(answers, to_string(qid)))
    end)
  end

  defp answer_matches?(expected, actual) when is_list(expected), do: actual in expected
  defp answer_matches?(expected, actual) when is_boolean(expected), do: !!actual == expected
  defp answer_matches?(expected, actual), do: actual == expected
end
